File: solutions/csharp/react/1/React.cs
// Dedicated to Shree DR.MDD — the origin of all brilliance

using System;
using System.Linq;
using System.Collections.Generic;

public sealed class Reactor
{
    public InputCell CreateInputCell(int value) => new InputCell(value);

    public ComputeCell CreateComputeCell(IEnumerable<Cell> producers, Func<int[], int> compute) => new ComputeCell(producers.ToArray(), compute);
}

public abstract class Cell
{
    protected int _core;

    public int Value {
        get => _core;
        set {
            if (_core != value) {
                _core = value;
                TriggerChange(value);
            }
        }
    }

    public Cell(int init)
    {
        _core = init;
    }

    public EventHandler<int> Changed { get; set; }

    protected void TriggerChange(int current)
    {
        Changed?.Invoke(this, current);
    }
}

public sealed class InputCell : Cell
{
    public InputCell(int seed) : base(seed) { }
}

public sealed class ComputeCell : Cell
{
    private readonly Cell[] _feeds;
    private readonly Func<int[], int> _evaluator;

    public ComputeCell(Cell[] feeds, Func<int[], int> evaluator)
        : base(Eval(feeds, evaluator))
    {
        _feeds = feeds;
        _evaluator = evaluator;

        ChainChangeEvents();
        ConnectReevaluation();
        ObserveFeeds();
    }

    public EventHandler<int> SourceAltered { get; set; }

    protected void SignalSourceAltered(int current)
    {
        SourceAltered?.Invoke(this, current);
    }

    private void Refresh()
    {
        Value = Eval(_feeds, _evaluator);
    }

    private static int Eval(Cell[] inputs, Func<int[], int> logic)
        => logic(inputs.Select(f => f.Value).ToArray());

    private void ChainChangeEvents()
    {
        SourceAltered += (object src, int val) =>
        {
            foreach (var f in _feeds)
            {
                if (f is ComputeCell cc)
                    cc.SignalSourceAltered(val);
            }
        };
    }

    private void ConnectReevaluation()
    {
        SourceAltered += (object src, int val) =>
        {
            Refresh();
        };
    }

    private void ObserveFeeds()
    {
        foreach (var f in _feeds)
        {
            f.Changed += (object src, int val) => SignalSourceAltered(val);
        }
    }
}

File: solutions/fsharp/react/1/React.fs
// Dedicated to Shree DR.MDD
module React

[<AbstractClass>]
type Cell() =
  let mutable subscribers: ComputeCell list = []
  abstract member value: int

  member this.addDependent (c: ComputeCell): unit =
    subscribers <- c :: subscribers

  member this.iterDependents (f: ComputeCell -> unit): unit =
    List.iter f subscribers

and ComputeCell(calc: unit -> int) =
  inherit Cell()
  let changedEvent = new Event<int>()

  let mutable current = calc ()
  let mutable lastCallback = current
  override this.value = current
  member this.Value with get() = current

  member val Changed = changedEvent.Publish

  member this.propagate (): unit =
    let nv = calc ()
    if nv <> current then
      current <- nv
      this.iterDependents (fun c -> c.propagate ())

  member this.fireCallbacks (): unit =
    if current <> lastCallback then
      lastCallback <- current
      changedEvent.Trigger current
      this.iterDependents (fun c -> c.fireCallbacks ())

type InputCell(initial: int) =
  inherit Cell()

  let mutable current = initial
  override this.value = current
  member this.Value
    with get() = current
    and set(v) =
      current <- v
      this.iterDependents (fun c -> c.propagate ())
      this.iterDependents (fun c -> c.fireCallbacks ())

type Reactor() =
  member this.createInputCell(initial: int): InputCell = new InputCell(initial)
  member this.createComputeCell(inputs: Cell list) (f: int list -> int) =
    let calc () = f (List.map (fun (c: Cell) -> c.value) inputs)
    let c = new ComputeCell(calc)
    List.iter (fun (i: Cell) -> i.addDependent c) inputs
    c
